src/LowerTriangularMatrices/lower_triangular_matrix.jl

```diff
@@ -162,166 +264,216 @@ function Base.copyto!(  L1::LowerTriangularMatrix{T},   # copy to L1
-    T = promote_type(eltype(L1), eltype(L2))
-    M = similar(L1, T)
-    @inbounds for i in eachindex(M, L1, L2)
-        M[i] = L1[i] + L2[i]
-    end
-    M
+function Base.:(+)(L1::LowerTriangularArray{T,N,ArrayType}, L2::LowerTriangularArray{S,N,ArrayType}) where {T,S,N,ArrayType}
+    @boundscheck (L1.m == L2.m) && (L1.n == L2.n) || throw(BoundsError)
+    LowerTriangularArray(L1.data + L2.data, L1.m, L1.n)
 end
 
-function Base.:(-)(L1::LowerTriangularMatrix, L2::LowerTriangularMatrix)
-    T = promote_type(eltype(L1), eltype(L2))
-    M = similar(L1, T)
-    @inbounds for i in eachindex(M, L1, L2)
-        M[i] = L1[i] - L2[i]
-    end
-    M
+function Base.:(-)(L1::LowerTriangularArray{T,N,ArrayType}, L2::LowerTriangularArray{S,N,ArrayType}) where {T,S,N,ArrayType}
+    @boundscheck (L1.m == L2.m) && (L1.n == L2.n) || throw(BoundsError)
+    LowerTriangularArray(L1.data - L2.data, L1.m, L1.n)
 end
 
-Base.:(-)(L::LowerTriangularMatrix) = LowerTriangularMatrix(-L.data, size(L)...)
-Base.prod(L::LowerTriangularMatrix{NF}) where NF = zero(NF)
+Base.:(-)(L::LowerTriangularArray) = LowerTriangularArray(-L.data, L.m, L.n)
+Base.prod(L::LowerTriangularArray{NF}) where NF = zero(NF)
 
 """
-    fill!(L::LowerTriangularMatrix, x)
-
+$(TYPEDSIGNATURES)
 Fills the elements of `L` with `x`. Faster than fill!(::AbstractArray, x)
 as only the non-zero elements in `L` are assigned with x."""
-function Base.fill!(L::LowerTriangularMatrix{T}, x) where T
+function Base.fill!(L::LowerTriangularArray{T}, x) where T
     xT = convert(T, x)
-    @inbounds for i in eachindex(L)
-        L[i] = xT
-    end
+    fill!(L.data, xT)
+
+    L
 end
 
-function scale!(L::LowerTriangularMatrix{T}, s::Number) where T
+function scale!(L::LowerTriangularArray{T}, s::Number) where T
     sT = convert(T, s)
-    @inbounds for i in eachindex(L)
-        L[i] *= sT
-    end
+    L.data .*= sT
+
     L
 end
 
-# Broadcast (more or less copied and adjusted from LinearAlgebra.jl)
-import Base: similar, copyto!
+# Broadcast CPU (more or less adjusted from LinearAlgebra.jl)
 import Base.Broadcast: BroadcastStyle, Broadcasted, DefaultArrayStyle
 import LinearAlgebra: isstructurepreserving, fzeropreserving
 
-struct LowerTriangularStyle <: Broadcast.AbstractArrayStyle{2} end
+struct LowerTriangularStyle{N} <: Broadcast.AbstractArrayStyle{N} end
 
-Base.BroadcastStyle(::Type{<:LowerTriangularMatrix}) = LowerTriangularStyle()
+Base.BroadcastStyle(::Type{LowerTriangularArray{T,N,ArrayType}}) where {T,N,ArrayType} = LowerTriangularStyle{N}()
 
-function Base.similar(bc::Broadcasted{LowerTriangularStyle}, ::Type{NF}) where NF
+function Base.similar(bc::Broadcasted{LowerTriangularStyle{N}}, ::Type{NF}) where {N,NF}
     inds = axes(bc)
     if isstructurepreserving(bc) || fzeropreserving(bc) 
-        return LowerTriangularMatrix{NF}(undef, inds[1][end], inds[2][end])
+        return LowerTriangularArray{NF,N,Array{NF}}(undef, [ind[end] for ind in inds]...)
```

Comment:
`[ind[end] for ind in axes(bc)]...` is just `size(bc)...` ???

Reply:
Also when are `isstructurepreserving(bc)` or `fzeropreserving(bc)` true?

Reply:
As far as I can see it they check if the broadcast result is a structured matrix like a lower triangular matrix is as well (imagine e.g an elemetwise multiplication with a full matrix with entries outside of the lower triangle). In 2D this was implemented in `Base.broadcast` for their lower triangular matrix. In higher dims I am not so sure anymore, so it's good that you point that out. We don't get any errors in the tests here though.

Reply:
Maybe we should at a test like 
```Julia 
@test typeof(L1 .+ L2) <: LowerTriangularArray 
```
I am not mistaking this should invoke the `similar` from the broadcast style.

Reply:
I had added some tests that check that the resulting type is still a `LowerTriangularArray` to avoid escaping while broadcasting!